### DISCUSSION TITLE
Automatic Stacked Layouts in preview 1.11.5_01

### DISCUSSION THREAD

**Discussion by dpolivaev** [freeplane/freeplane#1185] [Announcements]
Surprisingly I got a wish to develop a new automatic layout variant which I called "automatic stacked layouts".
They are available in the new preview at the bottom of the layout selector. It is sufficient to apply them to the root node and keep all other nodes just with just layout "Auto" as in the shown example, but you can generally use them at any node.

<img width="1195" alt="stacked layouts" src="https://github.com/freeplane/freeplane/assets/1558052/886ebd46-e68c-4158-9ba7-ae3f4ffc7101">

[Beatles.mm.zip](https://github.com/freeplane/freeplane/files/11593410/Beatles.mm.zip)

**Comment by dpolivaev** [comment]
I would like to ask people who tried it to share their experiences here.

**Comment by sifran-github** [comment]
OK I tried it just to give some feedback.

The layout changes the root layout and also the immediate the children layout.
if I take the central layout from your image, it will set:
- root -> "left to right, both sides, natural flow"
- immediate children on top -> "top to bottom, both sides, child nodes before parent"
- immediate children on bottom -> "top to bottom, both sides, child nodes after parent"

my comments:

1. If I'm going to use a similar layout then I personally think it looks better to use "center parent" instead of "natural flow" for the root item.
2. I do not really like the idea of having layout configuration that affect multiple nodes (current node AND children), just because it can add so much complexity... it can easily add 50 more layout configurations (root layout * different children layouts). I prefer to keep it to be configured manually, to reduce complexity. I understand that it might be nice to have a "one click to configure an entire map layout" but If you do decide to create a multi-level layouts... then I think it should be an addon or at least somehow visually separated from the regular layouts to make it very clear. but again, creating such multi-level layouts can create much more complexity and many more options.

**Reply by sifran-github** [reply]
in contrast to point 2, there is some advantage for multi-level layouts: you don't need to reconfigure a child node when you move it from top to bottom or vise versa.

but even if we do support multi-level layouts, I still prefer to configure each level manually instead of having one button for the entire map layout. It will be more generic and will allow more options. but its complex anyway.

what i mean is:
CURRENT NODE: [layout]

CHILDREN LEVEL 1 ON THE RIGHT: [layout]
CHILDREN LEVEL 1 ON THE LEFT: [layout]
CHILDREN LEVEL 1 ON THE TOP: [layout]
CHILDREN LEVEL 1 ON THE BOTTOM: [layout]

CHILDREN LEVEL 2 ON THE RIGHT: [layout] (note that the RIGHT is of LEVEL 1->LEVEL 2, not CURRENT->LEVEL1)
...

and so on.


maybe adding "templates" to this multi-level layout configuration can help, but anyway i want to be able to configure it manually.

**Reply by dpolivaev** [reply]
Technically it works as a multi-level layout only because it is handled by the map root node differently than for all other nodes. But it actually does not work across multiple levels. Given it is automatic I find it OK. Unfortunately, only a few people are likely to understand how it works unless they are given a template already using it.

**Reply by dpolivaev** [reply]
All you need to configure it is you assign it to the root node and keep all other nodes "AUTO"-layouts. You can also assign it as map default, it is effectively the same.

**Reply by sifran-github** [reply]
I'm torn between:
- keeping it simple and configurable, as it is. single level layouts.
- making it more complex with multi-level layouts.

If you are giving some option for multi level, even if its just from the root node, then why not building it in a configurable way? so users will be able to set, for example, the "center parent" instead of "natural flow" etc', or... that users will be able to configure multiple levels (root+level1+level2+level3) and save it as their template...?

if its not going to be configurable, and its going to be just "built in templates", then I guess there will be many many buttons in the GUI for every single combination... (or you will start with a few and later people will ask for more.... just like i'm asking for "central parent" and other people will want other things).

and also.. if its only for the root node, then this option should not exist for other nodes (currently you are able to choose this layout for every node). (Although... if we are going to allow templating... then it can theoretically support every node).

And I also still think that this should be separated from the other layouts. specially when there will be much more and not just 3. Depending on the implementation, it can be a just a simple space between them, or a completely different option with templates, level configuration and everything.

I hope it helps.

**Reply by dpolivaev** [reply]
Currently, I believe that I give you the best of both worlds: each layout can be applied to any node or style, and there are context-depending layouts contained in the last 3 rows and marked with a gear symbol. The 3 recently added layouts introduce no new concept.

If you have a concrete idea of what should be configurable and what the UI could look like you can share it.
